File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    public static void dfs(int v){
            visited[v]=1;
            
            for(int nextNode:graph[v]){
                if(visited[nextNode]==0){
                    visited[nextNode]=1;
                    dfs(nextNode);
                }
            }
    }
    static ArrayList<Integer>[] graph;
    static int[] visited;
    public int solution(int n, int[][] computers) {
        int cnt=0;
        visited=new int[n];
        // [] -> 리스트  따라서 리스트에 대한 배열이니까 이렇게가 맞음
        // 왼쪽이 안쪽, 오른쪽이 바깥쪽 자료구조(형)
        graph=new ArrayList[n];
        for(int i = 0;i<n;i++){
            graph[i]=new ArrayList<Integer>();
        }
        for(int i = 0;i<n;i++){
            for(int j = 0;j<n;j++){
                if(i!=j && computers[i][j]==1){
                    graph[i].add(j);
                    graph[j].add(i);
                }
            }
        }
        
        
        
        for(int i=0;i<n;i++){
            if(visited[i]==0){
                dfs(i);
                cnt++;
            }
        }
        
        return cnt;
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def solution(n, computers):
    visited=[0]*n
    cnt=0
    graph=[[] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if i!=j and computers[i][j]==1:
                graph[i].append(j)
                graph[j].append(i)
    
    def bfs(v):
        q=deque()
        q.append(v)
        visited[v]=1
        
        while q:
            node=q.popleft()
            for next_node in graph[node]:
                if visited[next_node]==0:
                    q.append(next_node)
                    visited[next_node]=1
    
    for i in range(n):
        if visited[i]==0:
            bfs(i)
            cnt+=1
    
    
    return cnt